graphql_test.go

```diff
@@ -55,19 +54,16 @@ type structFieldResolver struct {
 }
 
 type customDirectiveVisitor struct {
-	// Don't really want this... want to have the visitor be a standard struct with the directive (instance) properties
-	// In most cases, this will be created and bound to each relevant field in the schema once only
-	// Likely need some initialisation, similar to custom schema types to validate the instance on the schema?
-	directive *types.Directive
+	customAttribute *string
```

Comment:
It needs to be public if we want to access it from somewhere else. But there is another issue. This directive is going to be accessed from multiple goroutines and it is going to be added to multiple fields. So this is not thread-safe 🤔

Reply:
Technically speaking, GraphQL spec uses case sensitive names. However, I don't think its a good design to include the same arg with names only differing in casing. Therefore just like with method resolvers, args can be mapped to a public struct field (such that has its name starting with an uppercase letter).

Reply:
I absolutely love this idea of yours, it completely hides the `types` package and makes it user-friendly, however, imagine the following schema:
```graphql
schema {
	query: Query
}

type Query {
  field1: ResultOne! @hasRole(role: USER)
  field2: ResultTwo! @hasRole(role: EDITOR)
  field3: ResultThree! @hasRole(role: ADMIN)
}
...
```
Since you are passing a different role for each instance of the directive and each field can execute in parallel with the other fields we need to make sure that we have a per-resolver instance of the arguments. In the above example each there is only one directive for every field being resolved and it means that we might end up with concurrency issues. It would be nice if we received and argument to the ResolverVisitor, which is of type:
```go
// this is used only for directive arguments
args struct{
    CustomAttribute *string
}
```
in addition to the resolver input argument.

Reply:
Now I'm convinced that we need to use `packer.MakeStructPacker` to parse all the directive args and to make them struct fields. This logic can be reused 100%. Then we need an instance of each directive for each field. So each field will have `DirectiveVisitors` or `PackedDirectives`. Naming is to be decided. This way, we'll end up with the proper fields.

Reply:
Yeah that’s my intention here: an instance of the visitor for each directive declaration in the schema. So yes, where there are multiple of the same directive (in the schema as a whole, or on the same field), these would have separate instances of the visitor in memory, with each being setup with the args relevant to that instance.

As these are directives on the _field definition_ in the schema, a they can be instantiated once when parsing the schema, and reused across requests safely